Sources/Drill/Invocation.swift

```diff
@@ -46,61 +58,64 @@ public struct Invocation {
+
       switch options.mode {
       case .compile:
         fatalError("only Parse is implemented")
       case .dump(.tokens):
-        TokenDescriber.describe(tokens, to: &stdoutStream)
+        run(Passes.lex |> Pass(name: "Describe Tokens") { tokens, _ in
+          TokenDescriber.describe(tokens, to: &stdoutStream)
+        })
       case .dump(.file):
-        for token in tokens {
-          token.writeSourceText(to: &stdoutStream, includeImplicit: false)
-        }
+        run(Passes.lex |> Pass(name: "Reprint File") { tokens, _ -> Void in
+          for token in tokens {
+            token.writeSourceText(to: &stdoutStream, includeImplicit: false)
+          }
+        })
       case .dump(.shined):
-        let layoutTokens = layout(tokens)
-        for token in layoutTokens {
-          token.writeSourceText(to: &stdoutStream, includeImplicit: true)
-        }
+        run(shineFile |> Pass(name: "Dump Shined") { tokens, _ in
+          for token in tokens {
+            token.writeSourceText(to: &stdoutStream, includeImplicit: true)
+          }
+        })
       case .dump(.parse):
-        let layoutTokens = layout(tokens)
-        let parser = Parser(diagnosticEngine: engine, tokens: layoutTokens)
-        if let module = parser.parseTopLevelModule() {
+        run(parseFile |> Pass(name: "Dump Parsed") { module, _ in
           SyntaxDumper(stream: &stderrStream).dump(module)
-        }
+        })
       case .dump(.scopes):
-        let layoutTokens = layout(tokens)
-        let parser = Parser(diagnosticEngine: engine, tokens: layoutTokens)
-        let module = parser.parseTopLevelModule()!
-        let binder = NameBinding(topLevel: module, engine: engine)
-        print(binder.performScopeCheck(topLevel: module))
-      case .verify(.parse):
-        engine.unregister(printingConsumerToken)
-        let layoutTokens = layout(tokens)
-        let parser = Parser(diagnosticEngine: engine, tokens: layoutTokens)
-        _ = parser.parseTopLevelModule()
-        let verifier =
-          DiagnosticVerifier(input: contents,
-                             producedDiagnostics: engine.diagnostics)
-        verifier.verify()
-        return verifier.engine.hasErrors()
-      case .verify(.scopes):
+        run(scopeCheckFile |> Pass(name: "Dump Scopes") { module, _ in
+          print(module)
+        })
+      case .verify(let verification):
         engine.unregister(printingConsumerToken)
-        let layoutTokens = layout(tokens)
-        let parser = Parser(diagnosticEngine: engine, tokens: layoutTokens)
-        guard let module = parser.parseTopLevelModule() else {
-          return true
+        switch verification {
+        case .parse:
+          return run(makeVerifyPass(url: url, pass: parseFile,
+                                    context: context)) ?? true
+        case .scopes:
+          return run(makeVerifyPass(url: url, pass: scopeCheckFile,
+                                    context: context)) ?? true
```

Comment:
The magic constants make me nervous.  Perhaps `HadErrors` should be its own type that happens to look boolean at the end of the day.

Reply:
So, this function will never return `nil` for verify passes. I could just `!` them, but I definitely think a `typealias HadErrors = Bool` would help readability _a ton_.